metadata-ingestion/src/datahub/utilities/file_backed_collections.py

```diff
@@ -29,56 +36,70 @@
 _VT = TypeVar("_VT")
 
 
-@dataclass
-class _SqliteConnectionCache:
+class ConnectionWrapper:
     """
-    If you pass the same filename to multiple FileBacked* objects, they will
-    share the same underlying database connection. This also does ref counting
-    to drop the connection when appropriate.
+    Wraps a SQlite connection, allowing connection reuse across multiple FileBacked* objects.
 
-    It's necessary to use the same underlying connection because we're using
-    exclusive locking mode. It's useful to keep data from multiple FileBacked*
-    objects in the same SQLite database because it allows us to perform queries
-    across multiple tables.
+    This is necessary because we're using exclusive locking mode.
+    It's useful to keep data from multiple FileBacked* objects in the same
+    SQLite database because it allows us to perform queries across multiple tables.
 
-    This is used as a singleton class.
+    Also provides file cleanup using TemporaryDirectory, query debug logging, and
+    a context manager interface.
     """
 
-    _ref_count: Dict[pathlib.Path, int] = field(default_factory=dict)
-    _sqlite_connection_cache: Dict[pathlib.Path, sqlite3.Connection] = field(
-        default_factory=dict
-    )
-
-    def get_connection(self, filename: pathlib.Path) -> sqlite3.Connection:
-        if filename not in self._ref_count:
-            conn = sqlite3.connect(filename, isolation_level=None)
-
-            # These settings are optimized for performance.
-            # See https://www.sqlite.org/pragma.html for more information.
-            # Because we're only using these dbs to offload data from memory, we don't need
-            # to worry about data integrity too much.
-            conn.execute('PRAGMA locking_mode = "EXCLUSIVE"')
-            conn.execute('PRAGMA synchronous = "OFF"')
-            conn.execute('PRAGMA journal_mode = "MEMORY"')
-            conn.execute(f"PRAGMA journal_size_limit = {100 * 1024 * 1024}")  # 100MB
-
-            self._ref_count[filename] = 0
-            self._sqlite_connection_cache[filename] = conn
-
-        self._ref_count[filename] += 1
-        return self._sqlite_connection_cache[filename]
+    conn: sqlite3.Connection
+    filename: pathlib.Path
+    _directory: Optional[tempfile.TemporaryDirectory] = None
```

Comment:
Yup I'm bad

Reply:
Actually, it seems like this is acceptable:
https://peps.python.org/pep-0526/#class-and-instance-variable-annotations
https://docs.python.org/3/library/typing.html#typing.ClassVar

Which makes some sense because it'd be weird if adding a default to a type annotation converts it from an instance var to a class var. Functionally, this means you can change the default value of `_directory` in code by setting `ConnectionWrapper._directory = <some_value>`, which you couldn't if we just set this in `__init__`, but I guess we just shouldn't do that...

Overall don't really like this syntax with python typing, especially since mypy doesn't error if you reassign `ConnectionWrapper._directory`

Reply:
it might be appropriate to override the init then, but I'll leave that up to you